roaringbitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java

```diff
@@ -192,59 +192,69 @@ private static void rangeSanityCheck(final long rangeStart, final long rangeEnd)
 
   /**
    * Generate a new bitmap that has the same cardinality as x, but with
-   * all its values incremented by offset.
+   * all its values incremented by offset. The parameter offset can be
+   * negative.
+   *
+   * Values wrap around, being treated as unsigned integers. Thus, for
+   * example, if the bitmap contains the value 0, and you call this function
+   * with the offset -1, the resulting bitmap will contain the value
+   * 4294967295 (as an unsigned integer).
    *
    * @param x source bitmap
    * @param offset increment
    * @return a new bitmap
    */
   public static RoaringBitmap addOffset(final RoaringBitmap x, int offset) {
-    int container_offset = offset >>> 16;
-    int in_container_offset = offset % (1<<16);
+    int container_offset = Integer.divideUnsigned(offset, (1<<16));
+    int in_container_offset = Integer.remainderUnsigned(offset , (1<<16));
     if(in_container_offset == 0) {
-      RoaringBitmap answer = x.clone();
-      for(int pos = 0; pos < answer.highLowContainer.size(); pos++) {
-        int key = Util.toIntUnsigned(answer.highLowContainer.getKeyAtIndex(pos));
+      RoaringBitmap answer = new RoaringBitmap();
```

Comment:
We may preallocate answer's `RoaringArray` here as we know the size in advance.

Reply:
RoaringArray may still be resized in this case.
What do you think about 
```java
//RoaringBitmap.java
RoaringArray ra = x.highLowContainer.clone().iAddOffset(container_offset)
// or  x.highLowContainer.addOffset(container_offset)
return new RoaringBitmap(ra)
```
```java
//RoaringArray.java
public iAddOffset(int containerOffset) {
  for (int i=0;k<this.size;i++) {
    int key = ...;
    key+= containerOffset
    if((key >= 0) || (key <= 0xFFFF))  {
      keys[i]=key
    }
  }
}
```
This approach moves the responsibility of managing containers array into RoaringArray class. Clone gives us correctly preallocated arrays.

Reply:
But it's not correct. If I have the values 98303 and 98305 in a bitmap, if I shift left 2, I have the same number of keys. If I shift left by 65536, I have the same number of keys, and I can just subtract 1 from the existing key. If I shift left 32768, I now have 2 keys with 1 bit in each container. If the shift saturates rather than wraps, we could have far fewer keys than in the input.

Reply:
Not sure if I get your point @richardstartin . This branch of code shifts **only** by multiple of 2^16. This means we always have the same number of containers (excluding scenario when containers are discarded because of moving out of range).

Reply:
Sorry, I wrote that on a phone and didn’t scroll up to the of statement. I agree. My point was that you can’t do it in the general case, only when the offset divides 2^16. If you move it to roaring array you should move the divisibility check too.